settings/settings-url.php

```diff
@@ -242,91 +242,22 @@ public function form() {
 	}
 
 	/**
-	 * Sanitizes the settings before saving.
+	 * Prepare the received data before saving.
 	 *
-	 * @since 1.8
+	 * @since 3.7
 	 *
-	 * @param array $options Unsanitized options to save.
+	 * @param array $options Raw values to save.
 	 * @return array
 	 */
-	protected function update( $options ) {
-		$newoptions = array();
-
-		foreach ( array( 'force_lang', 'rewrite' ) as $key ) {
-			$newoptions[ $key ] = isset( $options[ $key ] ) ? (int) $options[ $key ] : 0;
-		}
-
-		if ( 3 == $options['force_lang'] && isset( $options['domains'] ) && is_array( $options['domains'] ) ) {
-			$newoptions['domains'] = array();
-			foreach ( $options['domains'] as $key => $domain ) {
-				if ( empty( $domain ) ) {
-					$lang = $this->model->get_language( $key );
-					pll_add_notice(
-						new WP_Error(
-							sprintf( 'pll_invalid_domain_%s', $key ),
-							sprintf(
-								/* translators: %s is a native language name */
-								__( 'Please enter a valid URL for %s.', 'polylang' ),
-								$lang->name
-							)
-						)
-					);
-				}
-				else {
-					$newoptions['domains'][ $key ] = esc_url_raw( trim( $domain ) );
-				}
-			}
-		}
-
-		foreach ( array( 'hide_default', 'redirect_lang' ) as $key ) {
-			$newoptions[ $key ] = isset( $options[ $key ] ) ? 1 : 0;
-		}
-
-		if ( 3 == $options['force_lang'] ) {
-			if ( ! class_exists( 'PLL_Xdata_Domain', true ) ) {
-				$newoptions['browser'] = 0;
-			}
-			$newoptions['hide_default'] = 0;
-		}
-
-		// Check if domains exist
-		if ( $newoptions['force_lang'] > 1 ) {
-			$this->check_domains( $newoptions );
-		}
-
-		return $newoptions; // Take care to return only validated options
-	}
-
-	/**
-	 * Check if subdomains or domains are accessible
-	 *
-	 * @since 1.8
-	 *
-	 * @param array $options new set of options to test
-	 * @return void
-	 */
-	protected function check_domains( $options ) {
-		$options = array_merge( $this->options, $options );
-		$model = new PLL_Model( $options );
-		$links_model = $model->get_links_model();
-		foreach ( $this->model->get_languages_list() as $lang ) {
-			$url = add_query_arg( 'deactivate-polylang', 1, $links_model->home_url( $lang ) );
-			// Don't redefine vip_safe_wp_remote_get() as it has not the same signature as wp_remote_get()
-			$response = function_exists( 'vip_safe_wp_remote_get' ) ? vip_safe_wp_remote_get( esc_url_raw( $url ) ) : wp_remote_get( esc_url_raw( $url ) );
-			$response_code = wp_remote_retrieve_response_code( $response );
+	protected function prepare_raw_data( array $options ): array {
+		$defaults = array(
+			'force_lang'    => 0,
+			'domains'       => array(),
+			'hide_default'  => 0,
+			'rewrite'       => 0,
+			'redirect_lang' => 0,
+		);
```

Comment:
I must say I'm not a big fan of this, it's a bit anti-pattern with what we've done so far regarding options classes and their encapsulation. 
What do you think?